system/HTTP/SiteURIFactory.php

```diff
@@ -0,0 +1,245 @@
+     */
+    public function createFromString(string $uri): SiteURI
+    {
+        // Validate URI
+        if (filter_var($uri, FILTER_VALIDATE_URL) === false) {
+            throw HTTPException::forUnableToParseURI($uri);
+        }
+
+        $parts = parse_url($uri);
+
+        if ($parts === false) {
+            throw HTTPException::forUnableToParseURI($uri);
+        }
+
+        $query = $fragment = '';
+        if (isset($parts['query'])) {
+            $query = '?' . $parts['query'];
+        }
+        if (isset($parts['fragment'])) {
+            $fragment = '#' . $parts['fragment'];
+        }
+
+        $relativePath = $parts['path'] . $query . $fragment;
+        $host         = $this->getValidHost($parts['host']);
+
+        return new SiteURI($this->appConfig, $relativePath, $host, $parts['scheme']);
+    }
+
+    /**
+     * Detects the current URI path relative to baseURL based on the URIProtocol
+     * Config setting.
+     *
+     * @param string $protocol URIProtocol
+     *
+     * @return string The route path
+     *
+     * @internal Used for testing purposes only.
+     */
+    public function detectRoutePath(string $protocol = ''): string
+    {
+        if ($protocol === '') {
+            $protocol = $this->appConfig->uriProtocol;
+        }
+
+        switch ($protocol) {
+            case 'REQUEST_URI':
+                $routePath = $this->parseRequestURI();
+                break;
+
+            case 'QUERY_STRING':
+                $routePath = $this->parseQueryString();
+                break;
+
+            case 'PATH_INFO':
+            default:
+                $routePath = $this->superglobals->server($protocol) ?? $this->parseRequestURI();
+                break;
+        }
+
+        return ($routePath === '/' || $routePath === '') ? '/' : ltrim($routePath, '/');
+    }
+
+    /**
+     * Will parse the REQUEST_URI and automatically detect the URI from it,
+     * fixing the query string if necessary.
+     *
+     * This method updates superglobal $_SERVER and $_GET.
+     *
+     * @return string The route path (before normalization).
+     */
+    private function parseRequestURI(): string
+    {
+        if (
+            $this->superglobals->server('REQUEST_URI') === null
+            || $this->superglobals->server('SCRIPT_NAME') === null
+        ) {
+            return '';
+        }
+
+        // parse_url() returns false if no host is present, but the path or query
+        // string contains a colon followed by a number. So we attach a dummy
+        // host since REQUEST_URI does not include the host. This allows us to
+        // parse out the query string and path.
+        $parts = parse_url('http://dummy' . $this->superglobals->server('REQUEST_URI'));
+        $query = $parts['query'] ?? '';
+        $path  = $parts['path'] ?? '';
+
+        // Strip the SCRIPT_NAME path from the URI
+        if (
+            $path !== '' && isset($this->superglobals->server('SCRIPT_NAME')[0])
```

Comment:
`isset($this->superglobals->server('SCRIPT_NAME')[0])`
What is this condition checked?
`$_SERVER['SCRIPT_NAME']` is a string. 

The condition comes from:
https://github.com/codeigniter4/CodeIgniter4/blob/6cedbe97a8863ccc6a3b4402fabe402ed24677e6/system/HTTP/IncomingRequest.php#L282-L286

Reply:
Does that end up being translated as this, because of string array access?
```php
$_SERVER['SCRIPT_NAME'] !== ''
```

Reply:
Yes. I assume this is unintended but that's the equivalence. 

https://3v4l.org/XC3NO
https://3v4l.org/hqebU

Reply:
Refactored. fde3216efdb84392aeb7c53df1f2df2c0a6bd74f